granulate_utils/metrics/sampler.py

```diff
@@ -263,12 +266,18 @@ def _init_collectors(self):
         This function fills in self._spark_samplers with the appropriate collectors.
         """
         if self._cluster_mode == SPARK_YARN_MODE:
-            self._collectors.append(YarnCollector(self._master_address, self._logger))
+            yarn_collector = YarnCollector(self._master_address, self._logger)
+            self._collectors.append(yarn_collector)
+            self._cluster_id = self._cluster_id or yarn_collector.cluster_id
 
-        # In Standalone and Mesos we'd use applications metrics
-        if self._cluster_mode in (SPARK_STANDALONE_MODE, SPARK_MESOS_MODE):
+        elif self._cluster_mode in (SPARK_STANDALONE_MODE, SPARK_MESOS_MODE):
+            # In Standalone and Mesos we'd use applications metrics
             self._applications_metrics = True
 
+            # Didn't research how to bring the actual cluster id, so use the master address
```

Comment:
Since we add cluster_id only for YARN, and for SA we remain with the master_address which is converted to "hostname" and is available in our queries - there's no purpose to add it as a label for SA and instead of having the rather patchy `_add_cluster_id_label`, you can add the YARN cluster_id inside `YarnCollector` to all metrics added by `YarnCollector`.

Reply:
note your suggestion does not support YARN with `applications_metrics` enabled

Reply:
I still think that having `Sampler` id is a good practice
If you find `_add_cluster_id_label` problematic or risking being high resource consumption, we might want to somehow transfer to the collectors `cluster_labels` which will be added by the collectors themselves while metrics are collected

Reply:
I still think that having `Sampler` id is a good practice
If you find `_add_cluster_id_label` problematic, we might want to somehow transfer to the collectors `cluster_labels` which will be added by the collectors themselves while metrics are collected